op-chain-ops/cmd/check-derivation/main.go

```diff
@@ -0,0 +1,432 @@
+package main
+
+import (
+	"context"
+	"crypto/ecdsa"
+	"errors"
+	"fmt"
+	"math/big"
+	"math/rand"
+	"os"
+	"time"
+
+	"github.com/ethereum-optimism/optimism/op-node/rollup"
+	"github.com/ethereum-optimism/optimism/op-service/eth"
+	"github.com/ethereum-optimism/optimism/op-service/retry"
+	"github.com/ethereum-optimism/optimism/op-service/testutils"
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/common/hexutil"
+	"github.com/ethereum/go-ethereum/core"
+	"github.com/ethereum/go-ethereum/core/types"
+	"github.com/ethereum/go-ethereum/crypto"
+	"github.com/ethereum/go-ethereum/ethclient"
+	"github.com/ethereum/go-ethereum/log"
+	"github.com/ethereum/go-ethereum/rpc"
+	"github.com/mattn/go-isatty"
+	"github.com/urfave/cli/v2"
+)
+
+func main() {
+	log.Root().SetHandler(log.StreamHandler(os.Stderr, log.TerminalFormat(isatty.IsTerminal(os.Stderr.Fd()))))
+
+	app := cli.NewApp()
+	app.Name = "check-derivation"
+	app.Usage = "Optimism derivation checker"
+	app.Commands = []*cli.Command{
+		{
+			Name:  "detect-l2-reorg",
+			Usage: "Detects unsafe block reorg",
+			Flags: []cli.Flag{
+				&cli.StringFlag{
+					Name:    "l2-rpc-url",
+					Value:   "http://127.0.0.1:9545",
+					Usage:   "L2 RPC URL",
+					EnvVars: []string{"L2_RPC_URL"},
+				},
+				&cli.IntFlag{
+					Name:  "polling-interval",
+					Value: 500,
+					Usage: "Polling interval (ms)",
+				},
+			},
+			Action: detectL2Reorg,
+		},
+		{
+			Name:  "check-consolidation",
+			Usage: "Checks consolidation",
+			Flags: []cli.Flag{
+				&cli.StringFlag{
+					Name:    "l2-rpc-url",
+					Value:   "http://127.0.0.1:9545",
+					Usage:   "L2 RPC URL",
+					EnvVars: []string{"L2_RPC_URL"},
+				},
+				&cli.IntFlag{
+					Name:  "polling-interval",
+					Value: 1000,
+					Usage: "Polling interval (ms)",
+				},
+				&cli.StringFlag{
+					Name:  "private-key",
+					Value: "ac0974bec39a17e36ba4a6b4d238ff944bacb478cbed5efcae784d7bf4f2ff80",
+					Usage: "Private key for signing L2 transactions. " +
+						"Default: devnet pre-funded account",
+				},
+				&cli.IntFlag{
+					Name:  "tx-count",
+					Value: 4,
+					Usage: "Number of transactions to send. Minimum value is 4 for checking every tx type.",
+				},
+				&cli.Uint64Flag{
+					Name:  "l2-chain-id",
+					Value: 901,
+					Usage: "L2 chain ID",
+				},
+			},
+			Action: checkConsolidation,
+		},
+	}
+
+	if err := app.Run(os.Args); err != nil {
+		log.Crit("error checking l2", "err", err)
+	}
+}
+
+type clients struct {
+	Client    *ethclient.Client
+	RpcClient *rpc.Client
+}
+
+func newClientsFromContext(ctx context.Context, cliCtx *cli.Context) (*clients, error) {
```

Comment:
We have a function that does this already here: https://github.com/ethereum-optimism/optimism/blob/e4de56021568cd12903541b4279baacf0662d8be/op-chain-ops/clients/clients.go#L24